Question 1a

In [16]:
import numpy as np
import matplotlib.pyplot as plt

def f(point):
    return np.exp(point)

def center(x, func, h):
    return (f(x+h) - f(x-h)) / (2*h)

h0 = center(1/2, f, 0.1)
print("h0 = ", h0)
h1 = center(1/2, f, 0.1/2)
print("h1 = ", h1)
h2 = center(1/2, f, 0.1/4)
print("h2 = ", h2)
h3 = center(1/2, f, 0.1/8)
print("h3 = ", h3)
h4 = center(1/2, f, 0.1/16)
print("h4 = ", h4)
h5 = center(1/2, f, 0.1/32)
print("h5 = ", h5)
h6 = center(1/2, f, 0.1/64)
print("h6 = ", h6)
h7 = center(1/2, f, 0.1/128)
print("h7 = ", h7)
h8 = center(1/2, f, 0.1/256)
print("h8 = ", h8)
h9 = center(1/2, f, 0.1/512)
print("h9 = ", h9)
h10 = center(1/2, f, 0.1/1024)
print("h10 = ", h10)


print("rate of convergence test 1 = ",(h1-h0) / (h2-h1))
print("rate of convergence test 2 = ",(h9-h8)/(h10-h9))
print("4 = 2^2 so we can verify that this converges at a quadratic rate ")

h0 =  1.6514705137461927
h1 =  1.6494083237722634
h2 =  1.648893017866171
h3 =  1.6487642064853247
h4 =  1.6487320045835219
h5 =  1.6487239541670817
h6 =  1.6487219415666488
h7 =  1.6487214384167714
h8 =  1.648721312629391
h9 =  1.6487212811824747
h10 =  1.6487212733215983
rate of convergence test 1 =  4.0018752930013735
rate of convergence test 2 =  4.00043387085111
4 = 2^2 so we can verify that this converges at a quadratic rate 


Question 1b

The optimal value of $h_0$ is $h_0 = \big(\frac{3eps|f(x_0)|}{M_3}\big)^{1/3} $, where eps $=2^{-52}$. Let $M_3 = 1.7$.

In [25]:
opt = (3*2**(-52)*f(1/2) / 1.7)**(1/3)
print("optimal h0 = ", opt)

test = center(1/2,f,opt)
print(test)

error = f(1/2) - test
print("error = ", error)

optimal h0 =  8.644766487225863e-06
1.6487212707222116
error =  -2.208344618281899e-11


Question 1c

Let $h=0.1$. Then we obtain the approximations $f'(x) = 1.6514705137461927$ for $h$ and $f'(x) = 1.6494083237722634$ for $h/2$. Using extrapolation, we can get fourth order approximation $N_2(h) = N_1(h/2)+ \frac{N_1(h/2)-N_1(h)}{2^2-1} = 1.6494083237722634 + \frac{1.6494083237722634 -1.6514705137461927 }{3} = 1.648720927$. Applying extrapolation again, we get $N_2(h/2) = N_1(h/4)+ \frac{N_1(h/4)-N_1(h/2)}{2^2-1} = 1.648893017866171+ \frac{1.648893017866171-1.6494083237722634}{3} =1.648721249 $

In [31]:
N1 = h1 + ((h1-h0)/3)
N2 = h2 + ((h2-h1)/3)
N4 = h3 +((h3-h2)/3)
print(N1,N2,N4)
rate = (N2-N1) / (N4-N2)
print(rate)
print("rate of convergence is 4")

1.648720927114287 1.6487212492308068 1.648721269358376
16.003746751957244
rate of convergence is 4
